src/message/fetchActions.js

```diff
@@ -222,6 +227,42 @@ const fetchTopMostNarrow = () => async (dispatch: Dispatch, getState: GetState)
   }
 };
 
+/**
+ * Calls an async function and if unsuccessful retries the call.
+ *
+ * If the function is an API call and the response has HTTP status code 4xx
+ * the error is considered unrecoverable and the exception is rethrown, to be
+ * handled further up in the call stack.
+ */
+export async function tryFetch<T>(func: () => Promise<T>): Promise<T> {
+  const MAX_TIME_MS: number = 60000;
+  const backoffMachine = new BackoffMachine();
+
+  const timeAtStart = Date.now();
+  // eslint-disable-next-line no-constant-condition
+  while (true) {
+    const msElapsed = Date.now() - timeAtStart;
+    // If the first attempt had a 60-second timeout, and it took 3
+    // seconds, then the second attempt should have a 57-second
+    // timeout. Etc.
+    const timeLimitMs = MAX_TIME_MS - msElapsed;
+    try {
+      return await promiseTimeout<T>(func(), timeLimitMs);
+    } catch (e) {
+      if (isClientError(e) || e instanceof TimeoutError) {
+        throw e;
+      }
+      await backoffMachine.wait();
```

Comment:
Ideally this would also be covered by the logic expressed by `timeLimitMs` above.

Reply:
Hmm, true—see one possible way this might look, at https://github.com/zulip/zulip-mobile/pull/4166#discussion_r442655214.